# MOCHA: Motif Orientation and Characterization for High-throughput Assays

This bioinformatic analysis will process enzyme profiling data, and evaluate the specificity of your enzyme.

Input data must be in a text file, with 1 substrate per line, and the substrates should have the same sequence lenght.

- For an example of a correctly formatted file, refer to: exampleData.txt

# Install Modules:

You will need to install the following Python modules:

    pip install Flask
    pip install logomaker
    pip install networkx
    pip install seaborn
    pip install wordcloud

# Running the Program:

To host the website on your device, run the "app.py" script.

- The code will generate an IP address and port number for the website. 

    - Once the IP has been an displayed, click on it and the application will open in your browser.

- Futher instructions will be displayed on the webpage.

For an example of a properly formated input file, see: exampleData.txt 

# Limitations:

This tool is only compatable with enzymes that interact with protein sequences.

All input sequences should have the same lenght.

Input sequences can only contain the standard 20 amino acids:

- The default list of AAs is defined by this line:

        AA = ['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']

    - If you need to expand this set of AAs, open functions.py and edit the list.

# Note:

Defending upon how long your extracted Motif is, the Bar Graphs might cut of part of the sequence when working with longer Motifs. Or they could have excessive space when displaying shorter Motifs.

- To adjust the spacing on the bottom of the figure, open the script: functions.py

- Find the line where this function the is defined: plotBarGraph
    
- Adjust the "bottom" parameter in the line: fig.subplots_adjust(bottom=0.11)

# Results:

The AAs in the input sequences are counted, and then the probability of finding a given AA at each position in the substrate is evaluated.

![image](https://github.com/user-attachments/assets/c93fe542-1d98-43fd-9649-4300de8d123b)

The probability distributions are used to evaluate the Positional Entropy (∆S).

- A high ∆S value indicates a high degree of specificity at that position.

<p align="center">
  <img src="https://github.com/user-attachments/assets/48639ab3-7e54-4123-998d-022d0dfcbd04" width="400">
</p>

![image](https://github.com/user-attachments/assets/48639ab3-7e54-4123-998d-022d0dfcbd04)

A Weblogo is generated by multiplying the ∆S value of a given position in the sequence by the AA probabilities at that position.

![image](https://github.com/user-attachments/assets/debb0208-9f08-4f91-8467-0487e05147d0)

We extract the Motif from the substrate sequence by removing the non-specific positions. 

- The non-specific positions have ∆S values that are below the user defined minimum ∆S value.

- The occurances of each Motif is counted and the user defined "N" number of the most common Motifs are displayed here:

![image](https://github.com/user-attachments/assets/21496236-7f4c-4c60-a7e5-4d54acf8e1b0)

We also evaluate the probability of finding a given Motif

![image](https://github.com/user-attachments/assets/ec6ea8dd-7bfe-47fa-9269-c59d0e9ecfe9)

The Motifs are also displayed in a Word Cloud

![image](https://github.com/user-attachments/assets/9808a006-d674-45f3-b9b6-1d9c2200e4b9)

Finally, we evaluate the specific combinations of AAs in the selected set of Motifs in a Suffix Tree.

- In this figure, the positons are arranged by their ∆S values, in decending order.

- The lines that connect the nodes indicate which AAs are combined within the most frequently observed Motifs in your dataset.

![image](https://github.com/user-attachments/assets/7ab6a873-4d90-4e6d-98df-2c54a070d8bb)

