# MOCHA: Motif Orientation and Characterization for High-throughput Assays

This bioinformatic analysis will process enzyme profiling data, and evaluate the specificity of your enzyme.

Input data must be in a text file, with 1 substrate per line, and the substrates should have the same sequence lenght.

- For an example of a correctly formatted file, refer to: exampleData.txt

# Install Modules:

You will need to install the following Python modules:

    pip install Flask
    pip install logomaker
    pip install networkx
    pip install seaborn
    pip install wordcloud

# Running the Program:

To host the website on your device, run the "app.py" script.

- The code will generate an IP address and port number for the website. 

    - Once the IP has been an displayed, click on it and the application will open in your browser.

- Futher instructions will be displayed on the webpage.

For an example of a properly formated input file, see: exampleData.txt 

# Limitations:

This tool is only compatable with enzymes that interact with protein sequences.

All input sequences should have the same lenght.

Input sequences can only contain the standard 20 amino acids:

- The default list of AAs is defined by this line:

        AA = ['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']

    - If you need to expand this set of AAs, open functions.py and edit the list.

# Note:

Defending upon how long your extracted Motif is, the Bar Graphs might cut of part of the sequence when working with longer Motifs. Or they could have excessive space when displaying shorter Motifs.

- To adjust the spacing on the bottom of the figure, open the script: functions.py

- Find the line where this function the is defined: plotBarGraph
    
- Adjust the "bottom" parameter in the line: fig.subplots_adjust(bottom=0.11)

# Results:

The AAs in the input sequences are counted, and then the probability of finding a given AA at each position in the substrate is evaluated.

<p align="center">
  <img src="https://github.com/user-attachments/assets/c93fe542-1d98-43fd-9649-4300de8d123b" width="700">
</p>

The probability distributions are used to evaluate the Positional Entropy (∆S).

<p align="center">
  ∆S = S<sub>Max</sub> - S<sub>Shannon</sub> = log<sub>2</sub>(20) - ∑(-prob<sub>AA</sub> × log<sub>2</sub>(prob<sub>AA</sub>))
</p>

- A high ∆S value indicates a high degree of specificity at that position.

<p align="center">
  <img src="https://github.com/user-attachments/assets/c6aebb9d-cfef-4036-9fa2-74a8e02677d1" width="700">
</p>

A Weblogo is generated by multiplying the ∆S value of a given position in the sequence by the AA probabilities at that position.

<p align="center">
  <img src="https://github.com/user-attachments/assets/0095e818-bc19-48da-a732-9da8b57d5bda" width="700">
</p>

We extract the Motif from the substrate sequence by removing the non-specific positions. 

- The non-specific positions have ∆S values that are below the user defined minimum ∆S value.

- The occurances of each Motif is counted and the user defined "N" number of the most common Motifs are displayed here:

<p align="center">
  <img src="https://github.com/user-attachments/assets/21496236-7f4c-4c60-a7e5-4d54acf8e1b0" width="700">
</p>

We also evaluate the probability of finding a given Motif

<p align="center">
  <img src="https://github.com/user-attachments/assets/ec6ea8dd-7bfe-47fa-9269-c59d0e9ecfe9" width="700">
</p>

The Motifs are also displayed in a Word Cloud

<p align="center">
  <img src="https://github.com/user-attachments/assets/9808a006-d674-45f3-b9b6-1d9c2200e4b9" width="700">
</p>

Finally, we evaluate the specific combinations of AAs in the selected set of Motifs in a Suffix Tree.

- In this figure, the positons are arranged by their ∆S values, in decending order.

- The lines that connect the nodes indicate which AAs are combined within the most frequently observed Motifs in your dataset.

<p align="center">
  <img src="https://github.com/user-attachments/assets/7ab6a873-4d90-4e6d-98df-2c54a070d8bb" width="700">
</p>

# Server Related Issues:

You may encounter an error similar to this:

     * Serving Flask app 'app'
     * Debug mode: on
    Address already in use
    Port 5000 is in use by another program. Either identify and stop that program, or start the server with a different port.

- This means that the port you are trying to use to host the website is not avalible.

    - We need to terminate the server before we can create a new website.

# Terminating The Server:

Leaving the process running can be a significant drain on your battery, so its best to turn things off after you are done.

This is also how to handle the previouly mentioned Server Related Issue.

The commands you enter into the terminal will depend upon your OS:

- MacOS:

    List open files on port 5000:

        lsof -i :5000

    - Inspect output, and look for "Python" in the "COMMAND" column:

            COMMAND     PID     NODE NAME
            Python    49231 ... 0t0  TCP localhost:commplex-main (LISTEN)
            Python    49288 ... 0t0  TCP localhost:commplex-main (LISTEN)
            Python    49288 ... 0t0  TCP localhost:commplex-main (LISTEN)

    Use the Process ID (49231, 49288) to terminate these processes:

         kill 49231 49288 

- Windows

    Inspect network connections on port 5000

        netstat -ano | findstr :5000
    
     - Inspect output, and look for the process ID (20352):

            TCP    127.0.0.1:5000         0.0.0.0:0              LISTENING       20352

    Lists all running tasks with matching process ID:

        tasklist /FI "PID eq 20352"

    - Inspect output, and look for "python.exe" in the "Image Name" column:
  
            Image Name                     PID Session Name        Session#    Mem Usage
            ========================= ======== ================ =========== ============
            python.exe                   20352 Console                    3     39,732 K

    Kill the process:

        taskkill /PID 20352 /F
